### PR TITLE
[PROF-3164] Add profiling to dd-trace-rb

### PR DESCRIPTION
**Pull request by ivoanjo**
Hello there! Let's start the conversation about getting this to master (it's been a while since #1016).

### Some notes:

I recognize this branch is quite big. Over the past few months, I've worked to extract fixes to common code as separate pull requests, so that although large, the modifications to existing files are minimal, and the bulk of the changes are just the additions of profiler, in separate files. This is the reason for only having 36 deletions in the current diff (as I'm opening this PR).

The feature itself is getting to a good shape. I still have a few more fixes to go in my list, but IMO those fixes can be done in master as well.

The branch has been rebased several times in the last months to fix any conflicts with master. These rebases were done with `--rebase-merges` and thus the history of PR merges I did since January 2021 has been kept (I think the previous history has been lost as some past rebases were not done with this option).

### My questions to the rest of the team are:

1. What's the best way to gate enablement of the feature, until we deem it ready for prime-time? How has this been done for other experimental features in the past?

    If there's no previous approach to follow, I was thinking we could hide the changes to the documentation (so as to not confuse users), and possibly we could refuse to start the profiler unless `ENV['DD_ALLOW_UNRELEASED'] == 'true'` or something similar.

2. How do we want to review this? I was thinking it'd be nice to do a pass over this whole codebase; would it make sense for us to schedule a fixed timebox and get together in a zoom and do a collaborative pass?

3. What else needs changing before this can go to master?´

### Thanks:

~~To my mom, for~~ @delner for A LOT of the AMAZING feature work (as betrayed in the commit logs); my work thus far has mostly been compatibility and bug fixing. Thanks David for doing an awesome job and handing me some solid underpinnings :)

### CODE REVIEW

**Comment by ivoanjo** [comment]
>  (I think the previous history has been lost as some past rebases were not done with this option).

I've been thinking that if there weren't that many rebases, I may be able to restore this history. I think that would be valuable (because it connects them to the PRs where they were discussed -- otherwise the context is lost, and GitHub doesn't see it either because the commit hashes changed).

Anyone else find this valuable, or is this just me nitpicking?

File: docs/GettingStarted.md [hunk 1]
```diff
@@ -2497,3 +2498,58 @@ However, additional instrumentation provided by Datadog can be activated alongsi
 | `OpenTracing::FORMAT_TEXT_MAP` | Yes        |                        |
 | `OpenTracing::FORMAT_RACK`     | Yes        | Because of the loss of resolution in the Rack format, please note that baggage items with names containing either upper case characters or `-` will be converted to lower case and `_` in a round-trip respectively. We recommend avoiding these characters or accommodating accordingly on the receiving end. |
 | `OpenTracing::FORMAT_BINARY`   | No         |                        |
+
+### Profiling
+
+*Currently available as BETA feature.*
```
Comment (marcotc) [inline docs/GettingStarted.md:2504]: I think one could read this as: _"I have a brand new, low criticality service that I want to add `ddtrace` to. I guess I'll test the new beta profiling feature"_.

If I'm not mistaken, just following these instructions will not display profiling results in users Datadog dashboard, is that right? If so, we should clarify this statement with something along the lines: "... if you'd like to try it, contact Datadog support at ... so we can enable the beta in your account."

File: lib/ddtrace/buffer.rb [hunk 2]
```diff
@@ -4,8 +4,8 @@
 # Trace buffer that accumulates traces for a consumer.
 # Consumption can happen from a different thread.
 module Datadog
-  # Buffer that stores objects. The buffer has a maximum size and when
-  # the buffer is full, a random object is discarded.
+  # Bounded buffer used to store profiling events.
```
Comment (marcotc) [inline lib/ddtrace/buffer.rb:7]: This comment doesn't seem correct, as the buffer is used by the tracer as well, not only profiling.

**Comment by marcotc** [comment]
Regarding commit history: they seem to be there (all the way back to a year ago), maybe not with all GitHub "metadata", but git-wise it looks legible.

**Comment by ivoanjo** [comment]
> Regarding commit history: they seem to be there (all the way back to a year ago), maybe not with all GitHub "metadata", but git-wise it looks legible.

The meh thing is that merge commits are missing, and so there's nothing to connect commits with the pull requests where they were discussed and reviewed. And since the commit hashes changed, GitHub loses the connection as well.

File: lib/ddtrace/buffer.rb (hunk 2 above)
Comment (ivoanjo) [reply lib/ddtrace/buffer.rb:7]: Good catch. I totally missed that this is something shared with other bits of ddtrace. Reverted in https://github.com/DataDog/dd-trace-rb/pull/1443 .

File: docs/GettingStarted.md (hunk 1 above)
Comment (ivoanjo) [reply docs/GettingStarted.md:2504]: I've just removed the docs inhttps://github.com/DataDog/dd-trace-rb/pull/1444 . This way we decouple "merging to master" with "actually releasing it".

**Comment by ivoanjo** [comment]
Hey @DataDog/apm-ruby , keep the comments coming! I'm serious in getting this merged ASAP, this is not just a placeholder while I continue to do work in the branch.

**Review by marcotc — DISMISSED** [review]
With https://github.com/DataDog/dd-trace-rb/pull/1444 and https://github.com/DataDog/dd-trace-rb/pull/1443 merged, I think this PR can be safely merged into `ddtrace`, without affecting its current users. 👍 

I understand that this is not the profiler's Final Form™️, but as it stands now, I'm confident we can continue development on master branch from this point on.

**Comment by ivoanjo** [comment]
Awesome, thanks Marco! 🙇 

I'll hold off on merging until CI in master is back to green (doesn't seem like a good idea to do such a big code dump with CI broken for unrelated reasons). Looking forward to be part of the cool kids club :)

**Comment by delner** [comment]
Once `master` is green, and this is rebased, I'll give this another quick pass and approve it then.

**Comment by delner** [comment]
CI is green, and looks good. Thanks for all your hard work @ivoanjo !

**Review by delner — APPROVED** [review]

**Pull request merged**